Harden ssh config defaults and remove old host key support

jeaye opened the pull request:
###### Motivation for this change
To better harden the SSH service by default. This was originally mentioned by @jpierre03 here: https://github.com/NixOS/nixpkgs/issues/20161#issuecomment-274335279 -- I'm just looking to get the ball rolling.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Things to note

This breaks backward compatibility with anyone using `supportOldHostKeys`. If people want that, I think `extraConfig` will do the trick.

---

jeaye committed:
nixos/ssh: Remove support for old host keys

jeaye committed:
nixos/ssh: Harden config defaults

Mic92 commented on nixos/modules/services/networking/ssh/sshd.nix:
```diff
@@ -363,14 +358,21 @@ in
 
-        # Allow DSA client keys for now. (These were deprecated
-        # in OpenSSH 7.0.)
-        PubkeyAcceptedKeyTypes +ssh-dss
```
This should be mentioned in the release notes. Otherwise people might get locked out.

jeaye replied:
Yep, this is important to mention in the release notes. I don't think it's been fully removed yet, just deprecated. 

> OpenSSH 7.0 and greater similarly disable the ssh-dss (DSA) public key algorithm. It too is weak and we recommend against its use. It can be re-enabled using the HostKeyAlgorithms configuration option:

https://www.openssh.com/legacy.html

Mic92 replied:
can be found here: `nixos/doc/manual/release-notes/rl-1803.xml`

jeaye replied:
```diff
diff --git a/nixos/doc/manual/release-notes/rl-1803.xml b/nixos/doc/manual/release-notes/rl-1803.xml
index c1fe692cee..8bc9cd9133 100644
--- a/nixos/doc/manual/release-notes/rl-1803.xml
+++ b/nixos/doc/manual/release-notes/rl-1803.xml
@@ -72,6 +72,16 @@ following incompatible changes:</para>
       <option>services.pgmanage</option>.
     </para>
   </listitem>
+  <listitem>
+    <para>
+      The OpenSSH service no longer enables support for DSA keys by default,
+      which could cause a system lock out. DSA support was
+      <link xlink:href="https://www.openssh.com/legacy.html">deprecated in OpenSSH 7.0</link>,
+      due to it being too weak. To re-enable support, add
+      <literal>PubkeyAcceptedKeyTypes +ssh-dss</literal> to the end of your
+      <option>services.openssh.extraConfig</option>.
+    </para>
+  </listitem>
 </itemizedlist>
 
 </section>
```

Look reasonable?

grahamc commented on nixos/modules/services/networking/ssh/sshd.nix:
```diff
@@ -54,8 +54,6 @@ let
     ));
   in listToAttrs (map mkAuthKeyFile usersWithKeys);
 
-  supportOldHostKeys = !versionAtLeast config.system.stateVersion "15.07";
```
In the release notes, after fixing their host keys, they can safely change their state version from anything prior to `17.03`, to `17.03`, as vetted here: https://search.nix.gsc.io/?q=stateVersion

jeaye committed:
nixos/ssh: Update 18.03 release notes

Since ssh-dss is no longer supported by default, users relying on those
keys for their login may be locked out. They should ideally use stronger
keys, but adding the support for ssh-dss back in can also be done
through extraConfig.

jeaye commented:
Anything more needed on this?

NeQuissimus commented:
Thank you for this, LGTM

jpierre03 commented:
@jeaye thanks.
A similar ssh client configuration can be acheived  https://github.com/NixOS/nixpkgs/issues/20161#issuecomment-274335279 and overriden if necessary with  `$HOME/.ssh/config` directives.

jeaye commented:
@jpierre03 Yes indeed. I'd like very much to see more things like this become (available) defaults in NixOS.
